File: Matrix.rb
# Written: 02/06/2015

class Matrix
  attr_accessor :values
  attr_accessor :m
  attr_accessor :n
  
  def initialize(values)
    @values = values
    @m = values.length
    @n = values[0].length
  end
  
  def display
    for row in 0...@m
      for col in 0...@n
        print @values[row][col].to_s
        print " "
      end
      print "\n"
    end
    print "\n"
  end
  
  def copy
    newValues = []
    for row in 0...@m
      newEntries = []
      for col in 0...@n
        newEntries.push(@values[row][col])
      end
      newValues.push(newEntries)
    end
    return Matrix.new(newValues)
  end
  
  def transpose
    rows = []
    for col in 0...@n
      entries = []
      for row in 0...@m
        entries.push(@values[row][col])
      end
      rows.push(entries)
    end
    return Matrix.new(rows)
  end
    
end

File: Haskell/Matrix.hs
module Matrix where

import Data.List

data Matrix a = Matrix { entries :: [[a]]
                     , m :: Int
                     , n :: Int
                     }

instance (Show a) => Show (Matrix a) where
    show (Matrix values _ _) = concat $ map (++ "\n") $ map (show) values
    
instance (Eq a) => Eq (Matrix a) where
    Matrix entries1 m1 n1 == Matrix entries2 m2 n2 =
        (entries1 == entries2) && (m1 == m2) && (n1 == n2)


-- Scale a matrix by a given number
scale :: (Num a) => a -> Matrix a -> Matrix a
scale c (Matrix values m' n') = let newEntries = map (map (*c)) values in
                                   Matrix newEntries m' n'

-- Add two matrices
add :: (Num a) => Matrix a -> Matrix a -> Maybe (Matrix a)
add (Matrix entries1 m' n') (Matrix entries2 m'' n'') =
    if (m' == m'') && (n' == n'')
    then let newEntries = zipWith (zipWith (+)) entries1 entries2 in
             Just $ Matrix newEntries m' n'
    else Nothing
    
-- Multiply two matrices
multiply :: (Num a) => Matrix a -> Matrix a -> Maybe (Matrix a)
multiply (Matrix entries1 m' n') (Matrix entries2 m'' n'') =
    if (n' == m'') -- In AB, the number of rows in A == the number of columns in B
    then let rows = entries1
             cols = transpose entries2
             productRow row = map (foldl (+) 0) $ zipWith (zipWith (*)) (replicate n'' row) cols
          in let productEntries = map (productRow) rows
             in Just $ Matrix productEntries m' n''
    else Nothing
    
